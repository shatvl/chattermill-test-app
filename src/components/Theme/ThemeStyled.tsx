import styled from 'styled-components';

export const ThemeWrapper = styled.dl`
  display: flex;
  align-items: center;
  margin: 0.75rem 0 0;
  padding-right: 1.5rem;
`;

export const ThemeEmoji = styled.dt`
  font-size: 1.25rem;
`;

export const ThemeDescription = styled.dd`
  margin: -0.25rem 0 0 0.75rem;
  color: ${({ theme }) => theme.colors.textSecondary};
`;
