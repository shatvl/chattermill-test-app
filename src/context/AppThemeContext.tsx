import React, { FC, createContext, useContext } from 'react';
import { DefaultTheme, ThemeProvider } from 'styled-components';

interface AppThemeContextProps {}

export interface AppTheme extends DefaultTheme {
  colors: {
    [key: string]: string;
  };
  zIndex: {
    [key: string]: number;
  };
  fonts: {
    header: string;
    main: string;
  };
  borderWidth: string;
}

// global theme for app, TODO: dark / light mode
export const getTheme = (): AppTheme => ({
  zIndex: {
    xs: 1,
    s: 2,
    m: 3,
    lg: 4,
    xlg: 5,
    xxlg: 6,
  },
  colors: {
    background: '#fff',
    primary: '#ff8489',
    primaryLight: 'rgba(255, 132, 136, .2)',
    primaryMedium: 'rgba(255, 132, 136, .5)',
    red: '#d65942',
    gray: '#c4c9cc',
    grayLight: 'e3e3e3',
    text: '#2e3939',
    textLighter: '#2e3939',
    textSecondary: '#707579',
    textMeta: '#bfbfbf',
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
    header: 'Hind Siliguri',
    main: 'Hind Siliguri',
  },
  borderWidth: '0.0625rem',
});

export const AppThemeContext = createContext<AppThemeContextProps>({});

export const AppThemeProvider: FC = ({ children }) => (
  //we could implement dark / light theme here, or user specific styles base on user session
  <AppThemeContext.Provider value={{}}>
    <ThemeProvider theme={getTheme()}>{children}</ThemeProvider>
  </AppThemeContext.Provider>
);

export const usePortalAppThemeContext = () => useContext(AppThemeContext);
