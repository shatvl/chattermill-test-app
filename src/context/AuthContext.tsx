import React, {
  FC, createContext, useContext, useState, useEffect,
} from 'react';
import jwt_decode from 'jwt-decode';
import { CHATTER_TOKEN_COOKIE, getCoockie } from '../services/cookie';
import { User } from '../models/User.model';

interface AuthContextProps {
  user: User;
  isAuthorized: () => boolean;
}

export const AuthContext = createContext<AuthContextProps>({
  user: null,
  isAuthorized: () => false,
});

export const AuthProvider: FC = ({ children }) => {
  const [user, setUser] = useState<User>();

  // decode token from params or try to retrieve from cookie
  const getUser = (token?: string): User => {
    const tok = token ?? getCoockie(CHATTER_TOKEN_COOKIE);

    if (!tok) {
      return null;
    }

    try {
      const decodedUser: { exp: number; id: string; orig_iat: number } = jwt_decode(tok);

      return decodedUser || null;
    } catch (error) {
      console.error(error);
      return null;
    }
  };

  const isAuthorized = () =>
    // check via getUser() method, because token could be expired and if users won't decoded navigate to login page
    !!getUser();
  useEffect(() => setUser(getUser()), []);

  return <AuthContext.Provider value={{ user, isAuthorized }}>{children}</AuthContext.Provider>;
};

export const useAuthContext = () => useContext(AuthContext);
