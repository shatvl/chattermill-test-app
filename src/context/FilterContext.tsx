import React, {
  FC, createContext, useContext, useState,
} from 'react';

interface FilterContextProps {
  themeId: number;
  setThemeId: (themeId: number) => void;
  clearFilters: () => void;
}

export const FilterContext = createContext<FilterContextProps>({
  themeId: null,
  setThemeId: () => {},
  clearFilters: () => {},
});

export const FilterProvider: FC = ({ children }) => {
  const [themeId, setThemeId] = useState<number>();

  const handleSelectThemeId = (themeId: number) => setThemeId(themeId);

  const handleClearFilters = () => setThemeId(null);

  return (
    <FilterContext.Provider value={{ themeId, setThemeId: handleSelectThemeId, clearFilters: handleClearFilters }}>
      {children}
    </FilterContext.Provider>
  );
};

export const useFilterContext = () => useContext(FilterContext);
