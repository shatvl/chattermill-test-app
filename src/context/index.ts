export { AppThemeContext } from './AppThemeContext';
export { usePortalAppThemeContext } from './AppThemeContext';
