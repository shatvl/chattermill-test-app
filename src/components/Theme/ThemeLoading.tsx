import React, { FC } from 'react';
import { LoadingBox } from '../Loading';
import { ThemeWrapper } from './ThemeStyled';

const ThemeLoading: FC = () => (
  <ThemeWrapper>
    <LoadingBox height={2} />
  </ThemeWrapper>
);

export default ThemeLoading;
