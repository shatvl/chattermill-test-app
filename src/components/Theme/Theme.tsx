import React, { FC, useEffect } from 'react';
import { Sentiment } from '../../models/ReviewTheme';
import { useThemeLazyQuery } from '../../services/theme/useThemeLazyQuery';
import ThemeLoading from './ThemeLoading';
import { ThemeDescription, ThemeEmoji, ThemeWrapper } from './ThemeStyled';

interface ThemeProps {
  id: number;
  sentiment: Sentiment;
}

const getThemeEmoji = (sentiment: Sentiment): JSX.Element => {
  switch (sentiment) {
    case -1:
      return <span>😡</span>;
    case 0:
      return <span>😐</span>;
    case 1:
      return <span>🤩</span>;
    default:
      return <span>😐</span>;
  }
};

const Theme: FC<ThemeProps> = ({ id, sentiment }) => {
  const [loadTheme, { data, loading, error }] = useThemeLazyQuery();

  useEffect(() => loadTheme(id), [id]);

  if (loading) {
    return <ThemeLoading />;
  }

  if (error) {
    return null;
  }

  if (!data) {
    return null;
  }

  return (
    <ThemeWrapper>
      <ThemeEmoji>{getThemeEmoji(sentiment)}</ThemeEmoji>
      {data.name && <ThemeDescription>{data.name}</ThemeDescription>}
    </ThemeWrapper>
  );
};

export default Theme;
