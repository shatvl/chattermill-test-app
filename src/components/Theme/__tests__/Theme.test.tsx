/* eslint-env jest */
import React from 'react';
import renderer from 'react-test-renderer';
import 'jest-styled-components';

import { waitRenderer } from '../../../testing/wait';
import { AppThemeProvider } from '../../../context/AppThemeContext';
import mockFetch from '../../../testing/mockFetch';
import Theme from '../Theme';

describe('renders theme', () => {
  it('theme snapshot', async () => {
    const tree = renderer.create(
      <AppThemeProvider>
        <Theme id={6374} sentiment={1}></Theme>
      </AppThemeProvider>,
    );

    mockFetch(200, { data: { id: 6374, name: 'General' } });

    await waitRenderer();

    expect(tree.toJSON()).toMatchSnapshot();
  });
});
