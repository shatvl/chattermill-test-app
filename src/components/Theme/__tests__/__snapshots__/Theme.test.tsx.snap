// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders theme theme snapshot 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  margin: 0.75rem 0 0;
  padding-right: 1.5rem;
}

.c1 {
  font-size: 1.25rem;
}

.c2 {
  margin: -0.25rem 0 0 0.75rem;
  color: #707579;
}

<dl
  className="c0"
>
  <dt
    className="c1"
  >
    <span>
      ðŸ¤©
    </span>
  </dt>
  <dd
    className="c2"
  >
    General
  </dd>
</dl>
`;
